sympy/core/numbers.py

```diff
@@ -438,6 +439,342 @@ def ilcm(*args):
+        raise ValueError('optimization for more than 3 values')
+    elif len(s) == 3 and optimize is None:
+        optimize = True
+
+    def reduce1(k, i):
+        if A[i] != 0:
+            q = iround(A[k], A[i])
+        else:
+            if 2*abs(la[k][i]) > D[i]:
+                q = iround(la[k][i], D[i])
+            else:
+                q = 0
+        if q != 0:
+            A[k] -= q*A[i]
+            for c in range(m):
+                B[k][c] -= q*B[i][c]
+            la[k][i] -= q*D[i]
+            for j in range(i):
+                la[k][j] -= q*la[i][j]
+
+    m = len(s)
+    B = [[0]*m for i in range(m)]
+    for i in range(m):
+        B[i][i] = 1
+    la = [[0]*i for i in range(m)]
+    D = [1]*(m + 1)  # m+1 referenced as D[-1]
+    A = list(s)
+    k = 1
+    while k < m:
+        j = k - 1
+        reduce1(k, j)
+        if A[j] != 0 or (A[j] == 0 and A[k] == 0 and
+                _d*(D[j - 1]*D[k] + la[k][j] ** 2) < _n*D[j]**2):
+            _swap(k, m, A, B, la, D)
+            if k > 1:
+                k = j
+        else:
+            for i in range(j - 1, -1, -1):
+                reduce1(k, i)
+            k += 1
+
+    # make gcd positive and update multipliers to be consistent
+    gcd = A[-1]
+    del A, la, D
+    if gcd < 0:
+        gcd = -gcd
+        B[-1] = [-i for i in B[-1]]
+
+    if optimize and len(s) == 3:
+        mag = lambda x: sum([i**2 for i in x])
+        add = lambda x, y: [i + j for i, j in zip(x, y)]
+        sub = lambda x, y: [i - j for i, j in zip(x, y)]
+        sign = lambda x: -1 if x < 0 else (1 if x > 0 else 0)
+        def best(x):
+            m = [mag(i) for i in x]
+            return x[m.index(min(m))]
+        key = lambda x: tuple([sign(i) for i in x])
+        b1, b2, b3 = B
+        opt = {
+            (1,1,1): lambda: best([b3, sub(b3, b2)]),
+            (1,-1,-1): lambda: best([b3, add(b3, b2)]),
+            (-1,1,-1): lambda: best([b3, add(b3, b2)]),
+            (-1,-1,1): lambda: best([b3, sub(b3, b2)]),
+            (-1,-1,-1): lambda: best([b3, sub(b3, b2), add(b3, add(b1, b2))]),
+            (1,1,-1): lambda: best([b3, sub(b3, b2), add(sub(b3, b1), b2)]),
+            (-1,1,1): lambda: best([b3, add(b3, b2), sub(sub(b3, b1), b2)]),
+            (1,-1,1): lambda: best([b3, add(b3, b2), sub(add(b3, b1), b2)])}
+        k = key(b3)
+        if k in opt:
+            b = opt[k]()
+            if b != B[-1]:
+                B[-1] = b
+
+    # make signs canonical for reduced basis
+    nneg = sum([1 for j in B[:-1] for i in j if i < 0])
+    npos = sum([1 for j in B[:-1] for i in j if i > 0])
+    if nneg > npos:
+        for j in range(len(B) - 1):
+            B[j] = [-i for i in B[j]]
+
+    return (gcd, B)
+
+
+def hnf(G, _n=3, _d=4):
+    """return ``n``, ``H``, ``P`` where ``n`` is the number of
+    iterations needed to find ``H`` and ``P`` such that ``H*G = P``
+    with ``H`` being in Hermite normal form and ``P`` being unimodular.
```

Comment:
Size of the multipliers(?). I don't understand enough to make a direct comparison, but consider the following example from the paper:
```python
>>> from sympy.matrices.normalforms import hermite_normal_form
>>> G = Matrix(10,10,lambda i,j:(i+1)**3*(j+1)**2+i+j+2)
>>> from sympy.core.numbers import hnf
>>> hermite_normal_form(G)
Matrix([
[2880, 9, 1432],
[2352, 8, 1169],
[1848, 7,  918],
[1380, 6,  685],
[ 960, 5,  476],
[ 600, 4,  297],
[ 312, 3,  154],
[ 108, 2,   53],
[   0, 1,    0],
[   0, 0,    1]])
>>> Matrix(hnf(G.tolist())[-2][:3]).T
Matrix([
[  1,  0,   0],
[  0,  1,   0],
[  7,  4,  12],
[ 22,  9,  36],
[ 45, 16,  72],
[ 76, 25, 120],
[115, 36, 180],
[162, 49, 252],
[217, 64, 336],
[280, 81, 432]])
```
Also, hnf gives the unimodular matrix in addition to the Hermitian normal form.

I tried another small random matrix and note this:
```python
>>> m
Matrix([
[98, 23, 61],
[71, 45, 78],
[20, 60, 16]])
>>> hermite_normal_form(m)
Matrix([
[21671, 4905, 19657],
[    0,    2,     1],
[    0,    0,     4]])
>>> Matrix(hnf(m.tolist())[-2])
Matrix([
[1, 0,   3871],
[0, 1,  36273],
[0, 0, 173368]])
```
Really, any more on this should be done by someone with more insight into these things. The Rosetta Code link has an extension of this code there. I started there but ended up being partially successful by going back to the original source.

Reply:
Is the `hermite_normal_form` matrix upside down?

Reply:
Maybe this could be made as a change to the existing `hermite_normal_form` function rather than as a new `hnf` function.

Reply:
It would be good to improve HNF in general because HNF is the standard method to solve systems of linear diophantine equations which it would be good if SymPy supported (#20682).

Reply:
The differences we're seeing between `hnf` and `hermite_normal_form` are all due to conflicting definitions. The HMM paper calls it "row Hermite normal form". Cohen prefers to work by columns.

To make `hermite_normal_form` give the same results as `hnf` on a given matrix `M`, do this:
1. Rotate `M` 90 degrees counterclockwise
2. Pass through `hermite_normal_form`
3. Rotate 90 degrees clockwise and flip top to bottom.

I'm seeing the makings of a little math paper for the American Math Monthly. : )
"Hermite Normal Form under Matrix symmetries of the Dihedral Group D4" LOL

In particular it's kind of fascinating to compare the middle columns of the two outputs in your first example: you got squares, and I got first powers! What's going on there?!

Reply:
On a more practical note, I suppose what this means for SymPy is that it should have:

1. At the core, the best/fastest HNF algorithm, which is probably this new one.
2. The ability to perform all necessary flips and rotations on matrices, to accommodate users working within any set of conventions as to what "HNF" means.
3. Maybe some convenience methods that put (1) and (2) together for known cases.

I don't know how many different sets of conventions are out there, but maybe for now it's enough to have functions called `row_hnf`, which could be a synonym for this new function, while `column_hnf` would recreate the existing `hermite_normal_form` by composing the right flips and rotations with `row_hnf`.

Reply:
> Maybe this could be made as a change to the existing hermite_normal_form function rather than as a new hnf function.

Yes, I would also tend to think this still belongs under `polys.matrices` instead of `core.numbers`.